modules/netaddress.go

```diff
@@ -26,27 +33,60 @@ func (na NetAddress) Port() string {
 	if err != nil {
 		return ""
 	}
+	if na.IsValid() != nil {
+		return ""
+	}
 	return port
 }
 
-// IsLoopback returns true for ip addresses that are on the same machine.
-func (na NetAddress) IsLoopback() bool {
-	if !na.IsValid() {
+// isLoopback returns true for ip addresses that are on the same machine.
+func (na NetAddress) isLoopback() bool {
+	host, _, err := net.SplitHostPort(string(na))
+	if err != nil {
 		return false
 	}
-
-	host := na.Host()
-	if ip := net.ParseIP(host); ip != nil && ip.IsLoopback() {
+	if host == "localhost" {
 		return true
 	}
-	if host == "localhost" || host == "::" || host == "0.0.0.0" {
+	if ip := net.ParseIP(host); ip != nil && ip.IsLoopback() {
 		return true
 	}
 	return false
 }
 
-// IsValid does nothing. Please ignore it.
-func (na NetAddress) IsValid() bool {
-	_, _, err := net.SplitHostPort(string(na))
-	return err == nil
+// IsValid returns nil if the NetAddress is valid. Valid is defined as being of
+// the form "host:port" such that "host" is not a loopback address (except in
+// testing) nor an unspecified IP address, and such that neither the host nor
+// port are blank, contain white space, or contain the null character.
+func (na NetAddress) IsValid() error {
+	host, port, err := net.SplitHostPort(string(na))
+	if err != nil {
+		return err
+	}
+
+	for _, runeValue := range host {
+		if unicode.IsSpace(runeValue) || runeValue == 0 {
+			return errors.New("host has invalid characters")
+		}
+	}
+	if host == "" {
+		return errors.New("host is blank")
+	}
+	if ip := net.ParseIP(host); ip != nil && ip.IsUnspecified() {
+		return errors.New("host is the unspecified address")
+	}
+
+	for _, runeValue := range port {
+		if unicode.IsSpace(runeValue) || runeValue == 0 {
+			return errors.New("port has invalid characters")
+		}
+	}
+	if port == "" {
+		return errors.New("port is blank")
+	}
```

Comment:
seems easier to me to call `strconv.Atoi` on the port.
```go
portInt, err := strconv.Atoi(port)
if err != nil || portInt < 0 || portInt > 65535 {
    return errors.New("invalid port")
}
```
this would catch unicode stuff too.